cocoa/src/toga_cocoa/window.py

```diff
@@ -288,11 +303,102 @@ def get_visible(self):
     # Window state
     ######################################################################
 
-    def set_full_screen(self, is_full_screen):
-        current_state = bool(self.native.styleMask & NSWindowStyleMask.FullScreen)
-        if is_full_screen != current_state:
+    def get_window_state(self):
+        if self.interface.content and bool(
+            self.interface.content._impl.native.isInFullScreenMode()
+        ):
+            return WindowState.PRESENTATION
+        elif bool(self.native.styleMask & NSWindowStyleMask.FullScreen):
+            return WindowState.FULLSCREEN
+        elif bool(self.native.isZoomed):
+            return WindowState.MAXIMIZED
+        elif bool(self.native.isMiniaturized):
+            return WindowState.MINIMIZED
+        else:
+            return WindowState.NORMAL
+
+    def set_window_state(self, state):
+        current_state = self.get_window_state()
+        if current_state == state:
+            return
+        if (
+            current_state != WindowState.NORMAL
+            and state != WindowState.NORMAL
+            and (getattr(self, "_pending_window_state_transition", None) is None)
```

Comment:
If we have the following:
```
Line 1: set_window_state(MINIMIZED)
Line 2: set_window_state(FULL_SCREEN)
Line 3: set_window_state(MAXIMIZED)
```

Then execution starts with:
```
Line 1: set_window_state(MINIMIZED)
```
Here, the window will go to `MINIMIZED` state, as:
https://github.com/beeware/toga/blob/d0dcb359d4e134f6c8412384ee0101c44535221a/cocoa/src/toga_cocoa/window.py#L338-L339
Then Executing:
```
Line 2: set_window_state(FULL_SCREEN)
```
will put `FULLSCREEN` as pending and call `set_window_state(NORMAL)`:
https://github.com/beeware/toga/blob/d0dcb359d4e134f6c8412384ee0101c44535221a/cocoa/src/toga_cocoa/window.py#L324-L333
Now in the `NORMAL` branch, after the window state is finished `NORMAL`-izing, it will complete the pending state i.e., `FULLSCREEN`:
https://github.com/beeware/toga/blob/d0dcb359d4e134f6c8412384ee0101c44535221a/cocoa/src/toga_cocoa/window.py#L397-L400
Then it will return to the main script, and execute:
```
Line 3: set_window_state(MAXIMIZED)
```
Now `MAXIMIZED` is put to pending and the window is set to `NORMAL`. After, the window is finished `NORMAL`-izing, the pending state i.e., `MAXIMIZED` is completed.

So, doing this:
```
Line 1: set_window_state(MINIMIZED)
Line 2: set_window_state(FULL_SCREEN)
Line 3: set_window_state(MAXIMIZED)
```
will have the following window state transitions: `MINIMIZED`->`NORMAL`->`FULLSCREEN`->`NORMAL`->`MAXIMIZED`.